File: app/src/main/java/com/akrivonos/a2chparser/fabrics/SubjectFactory.java
package com.akrivonos.a2chparser.fabrics;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.AsyncSubject;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.ReplaySubject;

public class SubjectFactory {

    public static <T> PublishSubject<T> createPublishSubject(Observer<? super T> observer) {
        final PublishSubject<T> subject = PublishSubject.create();
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer);
        return subject;
    }

    public static <T> ReplaySubject<T> createReplaySubject(Observer<? super T> observer) {
        final ReplaySubject<T> subject = ReplaySubject.create();
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer);
        return subject;
    }

    public static <T> BehaviorSubject<T> createBehaviorSubject(Observer<? super T> observer) {
        final BehaviorSubject<T> subject = BehaviorSubject.create();
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer);
        return subject;
    }

    public static <T> AsyncSubject<T> createAsyncSubject(Observer<? super T> observer) {
        final AsyncSubject<T> subject = AsyncSubject.create();
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer);
        return subject;
    }
}

File: app/src/main/java/com/akrivonos/a2chparser/fabrics/SubjectFactory.kt
package com.akrivonos.a2chparser.fabrics

import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.AsyncSubject
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.ReplaySubject

object SubjectFactory {

    fun <T> createPublishSubject(observer: Observer<T>): PublishSubject<T> {
        val subject = PublishSubject.create<T>()
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer)
        return subject
    }

    fun <T> createReplaySubject(observer: Observer<in T>): ReplaySubject<T> {
        val subject = ReplaySubject.create<T>()
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer)
        return subject
    }

    fun <T> createBehaviorSubject(observer: Observer<in T>): BehaviorSubject<T> {
        val subject = BehaviorSubject.create<T>()
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer)
        return subject
    }

    fun <T> createAsyncSubject(observer: Observer<in T>): AsyncSubject<T> {
        val subject = AsyncSubject.create<T>()
        subject
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer)
        return subject
    }
}
